### PR TITLE
create preallocated DataVolume for KubeVirt

### LINKED ISSUE #8491: Kubevirt Implement Data Volume cloning
We have to introduce a feature gate that will allow KubeVirt VMs on up scale to clone root disk images instead of downloading them from the internet.
That considers only VMs that come from the same Machine Deployment object.

It should work like described in those points:
1. User creates a new Machine Deployment
2. Machine Controller creates a new Virtual Machine with Data Volume that has an http source
3. Data Volume creates a new PVC and downloads the image over network
4. User scales up the machine
5. Machine Controller creates a new Virtual Machine with Data Volume that has a clone source that points to PVC from point no. 2 
6. New Data Volume clones a PVC

The full mechanism of cloning is described on the Red Hat blog (take a look at "Two ways to use DataVolume")
https://developers.redhat.com/blog/2020/11/18/using-multus-and-datavolume-in-kubevirt#using_datavolume_in_kubevirt

More about data volumes:
* import from http source https://github.com/kubevirt/containerized-data-importer/#import-from-url
* clone from another PVC (Data Volume creates a PVC) https://github.com/kubevirt/containerized-data-importer/#clone-another-pvc

This task requires a research to find a best way to locate PVC that comes from the first Virtual Machine.
Later, once we have a name of the PVC we should change Data Volume source from http to clone. https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/kubevirt/provider.go#L426-L430

Acceptance criteria:
* when scaling a Machine Deployment new VMs should clone images from the first created Data Volume (PVC) instead of downloading it again from the internet
* the mechanism should be enabled by a feature gate

### PR DESCRIPTION
**Pull request by sankalp-r**
**What does this PR do / Why do we need it**:
This PR introduces a new field  `AllocatedDataVolume` in `KubevirtCloudSpec` for DataVolume cloning.
**Does this PR close any issues?**:
Fixes #8491 

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:
UI for cluster creation with the KubeVirt provider needs to be updated to add this preallocated DataVolumes.

```release-note
None
```

Signed-off-by: Sankalp Rangare <sankalprangare786@gmail.com>

### CODE REVIEW

**Review by ahmedwaleedmalik — APPROVED** [review]
LGTM

/hold for @mfranczy to review

File: pkg/apis/kubermatic/v1/cluster.go [hunk 1]
```diff
@@ -916,6 +916,15 @@ type KubevirtCloudSpec struct {
 
 	Kubeconfig    string `json:"kubeconfig,omitempty"`
 	CSIKubeconfig string `json:"csiKubeconfig,omitempty"`
+
+	AllocatedDataVolume []AllocatedDataVolume `json:"allocatedDataVolume,omitempty"`
```
Comment (mfranczy) [inline pkg/apis/kubermatic/v1/cluster.go]: please change to `PreAllocatedDataVolumes` the same for json

File: pkg/provider/cloud/kubevirt/storage.go [hunk 2]
```diff
@@ -107,3 +113,57 @@ func reconcileCSIRoleRoleBinding(ctx context.Context, namespace string, client c
...
+func dataVolumeCreator(datavolume *cdiv1beta1.DataVolume) reconciling.NamedCDIv1beta1DataVolumeCreatorGetter {
+	return func() (name string, create reconciling.CDIv1beta1DataVolumeCreator) {
+		return datavolume.Name, func(dv *cdiv1beta1.DataVolume) (*cdiv1beta1.DataVolume, error) {
+			dv.Spec = datavolume.Spec
+			return dv, nil
+		}
+	}
+}
+
+func reconcileAllocatedDataVolume(ctx context.Context, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client) error {
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/storage.go]: please change as well to reconcilePreAllocatedDataVolume

**Review by mfranczy — CHANGES_REQUESTED** [review]
Looks good, but please change the naming

**Review by mfranczy — APPROVED** [review]
/approve
/lgtm

**Pull request merged**